exporter/elasticexporter/internal/translator/elastic/metrics.go

```diff
@@ -0,0 +1,219 @@
+				dropped++
+				continue
+			}
+			dps := doubleGauge.DataPoints()
+			for i := 0; i < dps.Len(); i++ {
+				dp := dps.At(i)
+				if dp.IsNil() {
+					dropped++
+					continue
+				}
+				metricsets.upsert(model.Metrics{
+					Timestamp: asTime(dp.Timestamp()),
+					Labels:    asStringMap(dp.LabelsMap()),
+					Samples: map[string]model.Metric{name: {
+						Value: dp.Value(),
+					}},
+				})
+			}
+		case pdata.MetricDataTypeIntSum:
+			intSum := metric.IntSum()
+			if intSum.IsNil() {
+				dropped++
+				continue
+			}
+			dps := intSum.DataPoints()
+			for i := 0; i < dps.Len(); i++ {
+				dp := dps.At(i)
+				if dp.IsNil() {
+					dropped++
+					continue
+				}
+				metricsets.upsert(model.Metrics{
+					Timestamp: asTime(dp.Timestamp()),
+					Labels:    asStringMap(dp.LabelsMap()),
+					Samples: map[string]model.Metric{name: {
+						Value: float64(dp.Value()),
+					}},
+				})
+			}
+		case pdata.MetricDataTypeDoubleSum:
+			doubleSum := metric.DoubleSum()
+			if doubleSum.IsNil() {
+				dropped++
+				continue
+			}
+			dps := doubleSum.DataPoints()
+			for i := 0; i < dps.Len(); i++ {
+				dp := dps.At(i)
+				if dp.IsNil() {
+					dropped++
+					continue
+				}
+				metricsets.upsert(model.Metrics{
+					Timestamp: asTime(dp.Timestamp()),
+					Labels:    asStringMap(dp.LabelsMap()),
+					Samples: map[string]model.Metric{name: {
+						Value: dp.Value(),
+					}},
+				})
+			}
+		case pdata.MetricDataTypeIntHistogram:
+			// TODO(axw) requires https://github.com/elastic/apm-server/issues/3195
+			intHistogram := metric.IntHistogram()
+			if intHistogram.IsNil() {
+				dropped++
+				continue
+			}
+			dropped += intHistogram.DataPoints().Len()
+		case pdata.MetricDataTypeDoubleHistogram:
+			// TODO(axw) requires https://github.com/elastic/apm-server/issues/3195
+			doubleHistogram := metric.DoubleHistogram()
+			if doubleHistogram.IsNil() {
+				dropped++
+				continue
+			}
+			dropped += doubleHistogram.DataPoints().Len()
+		default:
+			// Unknown type, so just increment dropped by 1 as a best effort.
+			dropped++
+		}
+	}
+	for _, metricset := range metricsets {
+		w.RawString(`{"metricset":`)
+		if err := metricset.MarshalFastJSON(w); err != nil {
+			return dropped, err
+		}
+		w.RawString("}\n")
+	}
+	return dropped, nil
+}
+
+func asTime(in pdata.TimestampUnixNano) model.Time {
+	return model.Time(time.Unix(0, int64(in)))
+}
+
+func asStringMap(in pdata.StringMap) model.StringMap {
+	var out model.StringMap
+	in.Sort()
+	in.ForEach(func(k string, v pdata.StringValue) {
+		out = append(out, model.StringMapItem{
+			Key:   k,
+			Value: v.Value(),
+		})
+	})
+	return out
+}
+
+type metricsets []model.Metrics
+
+func (ms *metricsets) upsert(m model.Metrics) {
+	i := ms.search(m)
+	if i < len(*ms) && compareMetricsets((*ms)[i], m) == 0 {
+		existing := (*ms)[i]
+		for k, v := range m.Samples {
+			existing.Samples[k] = v
+		}
+	} else {
+		head := (*ms)[:i]
+		tail := append([]model.Metrics{m}, (*ms)[i:]...)
```

Comment:
Just curious, do we want to place a new metricset at the same index when `compareMetricsets((*ms)[i], m) == -1` and `compareMetricsets((*ms)[i], m) == 1`?

Reply:
I assume you're referring to the `compareMetricsets((*ms)[i], m) == 0` call above?

`search` (which just wraps https://golang.org/pkg/sort/#Search) returns the smallest index `i` at which `ms[i] >= m`. The additional call to `compareMetricsets` above is only to check for equality. If they're not equal, then we know the metric set doesn't already exist, and we must insert and not update.